main/src/ca/uwaterloo/flix/language/phase/optimizer/Inliner.scala

```diff
@@ -0,0 +1,214 @@
+    * Checks if `occur` is Dead.
+    */
+  private def isDead(occur: OccurrenceAst.Occur): Boolean = occur match {
+    case Dead => true
+    case _ => false
+  }
+
+  /**
+    * Checks if `occur` is Once and `eff` is Pure
+    */
+  private def isUsedOnceAndPure(occur: OccurrenceAst.Occur, eff0: Type): Boolean = occur match {
+    case Once => isPure(eff0)
+    case _ => false
+  }
+
+  /**
+    * Checks if `exp0` is trivial and `eff` is pure
+    */
+  private def isTrivialAndPure(exp0: Expr, eff0: Type): Boolean = {
+    isPure(eff0) && isTrivialExp(exp0)
+  }
+
+  /**
+    * Checks if `occur` is dead and `exp` is pure.
+    */
+  private def isDeadAndPure(occur: OccurrenceAst.Occur, eff0: Type): Boolean = occur match {
+    case Dead => isPure(eff0)
+    case _ => false
+  }
+
+  /** Returns a canonical effect type equivalent to `eff` */
+  private def canonicalEffect(eff: Type): Type = {
+    evalToType(eval(eff), eff.loc)
+  }
+
+  /** Evaluates a ground, simplified effect type */
+  private def eval(eff: Type): CofiniteSet[Symbol.EffectSym] = eff match {
+    case Type.Univ => CofiniteSet.universe
+    case Type.Pure => CofiniteSet.empty
+    case Type.Cst(TypeConstructor.Effect(sym), _) =>
+      CofiniteSet.mkSet(sym)
+    case Type.Apply(Type.Cst(TypeConstructor.Complement, _), y, _) =>
+      CofiniteSet.complement(eval(y))
+    case Type.Apply(Type.Apply(Type.Cst(TypeConstructor.Union, _), x, _), y, _) =>
+      CofiniteSet.union(eval(x), eval(y))
+    case Type.Apply(Type.Apply(Type.Cst(TypeConstructor.Intersection, _), x, _), y, _) =>
+      CofiniteSet.intersection(eval(x), eval(y))
+    case Type.Apply(Type.Apply(Type.Cst(TypeConstructor.Difference, _), x, _), y, _) =>
+      CofiniteSet.difference(eval(x), eval(y))
+    case Type.Apply(Type.Apply(Type.Cst(TypeConstructor.SymmetricDiff, _), x, _), y, _) =>
+      CofiniteSet.xor(eval(x), eval(y))
+    case other => throw InternalCompilerException(s"Unexpected effect $other", other.loc)
+  }
+
+  /** Returns the [[Type]] representation of `set` with `loc`. */
+  private def evalToType(set: CofiniteSet[Symbol.EffectSym], loc: SourceLocation): Type = set match {
+    case CofiniteSet.Set(s) => Type.mkUnion(s.toList.map(sym => Type.Cst(TypeConstructor.Effect(sym), loc)), loc)
+    case CofiniteSet.Compl(s) => Type.mkComplement(Type.mkUnion(s.toList.map(sym => Type.Cst(TypeConstructor.Effect(sym), loc)), loc), loc)
+  }
+
+
+  /**
+    * Returns `true` if `exp0` is considered a trivial expression.
+    *
+    * An expression is trivial if:
+    * It is either a literal (float, string, int, bool, unit), or it is a variable.
+    *
+    * A pure and trivial expression can always be inlined even without duplicating work.
+    */
+  private def isTrivialExp(exp0: Expr): Boolean = exp0 match {
+    case Expr.Cst(_, _, _) => true
+    case Expr.Var(_, _, _) => true
+    case Expr.ApplyAtomic(AtomicOp.Unary(_), exps, _, _, _) => exps.forall(isTrivialExp) // TODO: Do constant folding?
```

Comment:
I would not include unary and binary here, only tag, tuple, and other constructors.

Also no constant folding. We are not doing that today.

Reply:
Relevant constructors seem to be:

- RecordExtend
- Index
- `Is`? (what is this even?)
- StructNew
- StructPut
- InstanceOf
- ArrayNew
- ArrayLit
- ArrayLoad
- ArrayStore

Thoughts?

Reply:
I think maybe Unary and Binary are needed to specialize higher-order functions